### PR TITLE
Add AbstractPeriodicJob to the bg job lib

### PR DESCRIPTION
**Pull request by kibertoad**
## Changes

Make the AbstractPeriodicJob reusable across libraries

## Checklist

- [X] Apply one of following labels; `major`, `minor`, `patch` or `skip-release`
- [X] I've updated the documentation, or no changes were necessary
- [X] I've updated the tests, or no changes were necessary

### CODE REVIEW

File: packages/app/healthcheck-utils/src/AbstractPeriodicJob.ts [hunk 1]
```diff
@@ -113,7 +113,7 @@ export abstract class AbstractPeriodicJob {
       shouldLogExecution: false,
       ...options,
       singleConsumerMode: {
-        enabled: true,
+        enabled: false,
```
Comment (kibertoad) [inline packages/app/healthcheck-utils/src/AbstractPeriodicJob.ts:116]: disabled by default seems to be a less surprising setup for me

File: packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+import { resolveGlobalErrorLogObject } from '@lokalise/node-core'
+import type Redis from 'ioredis'
+import { stdSerializers } from 'pino'
+import type { LockOptions } from 'redis-semaphore'
+import { Mutex } from 'redis-semaphore'
+import type { LockLostCallback } from 'redis-semaphore'
+import type { ToadScheduler } from 'toad-scheduler'
+import { AsyncTask, SimpleIntervalJob } from 'toad-scheduler'
+
+const DEFAULT_JOB_INTERVAL = 60000
```
Comment (CarlosGamero) [inline packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts:18]: Does it mean that by default job will be executed once per minute? could we put a comment to clarify it 🙏

File: packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+const DEFAULT_EXCLUSIVE_LOCK_SUFFIX = 'EXCLUSIVE'
+
+export type BackgroundJobConfiguration = {
```
Comment (CarlosGamero) [inline packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts:21]: Nit: To make this file easier to read, could we extract types a separate file

File: packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+    /**
+     * Lock will be reset to this value after success, so that other node could potentially acquire the lock after it expires, but in order to prevent immediate acquire
+     */
+    lockTimeoutAfterSuccess?: number
+  }
+  /**
+   * If true, the job will log when it starts and finishes.
+   */
+  shouldLogExecution?: boolean
```
Comment (CarlosGamero) [inline packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts:50]: Shouldn't we always log start and finish?

File: packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+export type PeriodicJobDependencies = {
+  redis: Redis
+  logger: CommonLogger
+  transactionObservabilityManager: TransactionObservabilityManager
+  errorReporter: ErrorReporter
+  scheduler: ToadScheduler
+}
+
+export abstract class AbstractPeriodicJob {
+  public readonly options: Required<BackgroundJobConfiguration>
```
Comment (CarlosGamero) [inline packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts:97]: Should options be public?

File: packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+  }
+
+  public async dispose() {
+    this.scheduler.stopById(this.options.jobId)
+    await this.singleConsumerLock?.release()
+  }
+
+  protected abstract processInternal(executionUuid: string): Promise<unknown>
+
+  public async process(executorId: string) {
```
Comment (CarlosGamero) [inline packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts:158]: On `AbstractBackgroundJobProcessor` we use these names but in the other direction, should we follow the same approach? (`process` for the abstract method and `processInternal` for the private one)

File: packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts (hunk 2 above)
Comment (kibertoad) [reply packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts]: after thinking about it some more, I've removed the default, as every job is going to be different, having arbitrary default is going to be more harmful than useful

File: packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts (hunk 6 above)
Comment (kibertoad) [reply packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts]: not sure. note that there needs to be a public entry point that toad-scheduler task can invoke. would you call that `processInternal`?

File: packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts (hunk 2 above)
Comment (CarlosGamero) [reply packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts]: I agree with that! thanks Igor 🙏

File: packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts (hunk 4 above)
Comment (CarlosGamero) [reply packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts]: I see and it makes sense, but if I am not wrong it is `false` by default, I would expect to have it `true` but no strong opinion from my side

File: packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts (hunk 6 above)
Comment (CarlosGamero) [reply packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts]: Oh I see, sorry for the misunderstanding, wondering if we can make the method private and `createTask`  receive the method to call, something like:
```
createTask(logger: CommonLogger, jobId: string, method: (execId: string) => Promise<void>): AsyncTask
```

But I see createTask is exposed, so maybe it is not a good idea 🤔

Comment (kibertoad) [reply packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts]: yeah, I would probably keep it as-is, not sure that tinkering with this part would decrease complexity significantly

Comment (CarlosGamero) [reply packages/app/background-jobs-common/src/periodic-jobs/AbstractPeriodicJob.ts]: Yeah, I am fine with it, my only point was not allowing to directly call that method or override it

**Review by CarlosGamero — APPROVED** [review]

**Pull request merged**